src/pymor/algorithms/cholesky_qr.py

```diff
@@ -0,0 +1,144 @@
+        in :cite:`FKNYY20`) which is stable but leads to poor orthogonality in general, i.e.
+
+        .. math::
+            \lVert Q^TQ - I \rVert_2 < 2,
+
+        (see Lemma 3.1 in :cite:`FKNYY20`).
+
+    Parameters
+    ----------
+    A
+        The |VectorArray| which is to be orthonormalized.
+    product
+        The inner product |Operator| w.r.t. which to orthonormalize.
+        If `None`, the Euclidean product is used.
+    return_R
+        If `True`, the R matrix from QR decomposition is returned.
+    maxiter
+        Maximum number of iterations. Defaults to 3.
+    offset
+        Assume that the first `offset` vectors are already orthonormal and apply the
+        algorithm for the vectors starting at `offset + 1`.
+    orth_tol
+        If not `None`, check if the resulting |VectorArray| is really orthornormal and
+        repeat the algorithm until the check passes or `maxiter` is reached.
+    check_finite
+        This argument is passed down to |SciPy linalg| functions. Disabling may give a
+        performance gain, but may result in problems (crashes, non-termination) if the
+        inputs do contain infinities or NaNs.
+    copy
+        If `True`, create a copy of `A` instead of modifying `A` in-place.
+
+    Returns
+    -------
+    Q
+        The orthonormalized |VectorArray|.
+    R
+        The upper-triangular/trapezoidal matrix (if `compute_R` is `True`).
+    """
+    assert 0 <= offset < len(A)
+    assert 0 < maxiter
+    assert orth_tol is None or 0 < orth_tol
+
+    logger = getLogger('pymor.algorithms.cholesky_qr.shifted_chol_qr')
+
+    if maxiter == 1:
+        logger.warning('Single iteration shifted CholeskyQR can lead to poor orthogonality!')
+
+    if copy:
+        A = A.copy()
+
+    iter = 1
+    shift = None
+    while iter <= maxiter:
+        with logger.block(f'Iteration {iter}'):
+            if orth_tol is None or iter == 1:
+                # compute only the necessary parts of the Gramian
+                B, X = np.split(A[offset:].inner(A, product=product), [offset], axis=1)
+
+            # This will compute the Cholesky factor of the lower right block
+            # and keep applying shifts if it breaks down.
+            while True:
+                try:
+                    Rx = spla.cholesky(X, overwrite_a=True, check_finite=check_finite)
+                    break
+                except spla.LinAlgError:
+                    logger.warning('Cholesky factorization broke down! Matrix is ill-conditioned.')
+                    if shift is None:
+                        m, n = A.dim, len(A[offset:])
+                        if product is None:
+                            shift = m*n+n*(n+1)
+                            XX = X
+                        else:
+                            from pymor.algorithms.eigs import eigs
+                            shift = 2*m*np.sqrt(m*n)+n*(n+1)*np.sqrt(np.abs(eigs(product, k=1)[0][0]))
```

Comment:
Are the mathematics of this case clear? Is it covered in the original paper?

Reply:
Yes. See Theorem 5.1